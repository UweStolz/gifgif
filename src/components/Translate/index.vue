<template>
  <v-container fluid>
    <v-row
      align="start"
      justify="center"
    >
      <v-card
        flat
        tile
        :width="$vuetify.breakpoint.xsOnly ? '90%' : '80%'"
      >
        <v-slider
          v-model="weirdnessSlider"
          label="Weirdness factor"
          style="margin-top: 30px;"
          track-fill-color="red darken-3"
          track-color="black"
          thumb-color="red"
          min="1"
          max="10"
          thumb-label="always"
          ticks="always"
        />
        <v-form
          :value="isInputValid"
          @submit="getTranslatedGif"
        >
          <v-text-field
            v-model="inputValue"
            :error="hasError"
            :success="wasSuccessful"
            :error-messages="hasError ? 'Oh no, something went wrong!' : ''"
            :prepend-icon="icons.mdiMagnify"
            :prepend-inner-icon="icons.mdiEmoticonOutline"
            maxlength="25"
            color="black"
            clearable
            counter
            placeholder="Enter a search phrase"
            @click:prepend-inner="openEmojiMart"
            @click:clear="clearTextField"
          />
        </v-form>
        <Picker
          v-if="showPicker && !$vuetify.breakpoint.xsOnly"
          :style="{ position: 'absolute', bottom: '-345px', right: '362px' }"
          set="emojione"
          emoji=""
          :show-preview="false"
          :show-skin-tones="false"
          :sheet-size="32"
          title="Pick your emojiâ€¦"
          @select="addEmoji"
        />
        <v-row
          align="start"
          justify="center"
        >
          <v-img
            v-if="translatedGif"
            :src="translatedGif"
            :lazy-src="translatedGif"
            contain
            max-height="500"
            max-width="600"
            height="100%"
            width="100%"
          >
            <template v-slot:placeholder>
              <v-progress-linear
                indeterminate
                color="red"
              />
            </template>
          </v-img>
        </v-row>
        <v-card-actions>
          <v-row justify="center">
            <v-item-group v-if="translatedGif">
              <v-btn
                :disabled="rating === 0"
                large
                icon
                @click="removeGif"
              >
                <v-icon
                  id="trashIcon"
                  large
                >
                  {{ icons.mdiTrashCan }}
                </v-icon>
              </v-btn>
              <v-btn
                large
                icon
                @click="saveImage"
              >
                <v-icon
                  id="downloadIcon"
                  large
                >
                  {{ icons.mdiDownload }}
                </v-icon>
              </v-btn>
              <v-rating
                v-model="rating"
                length="5"
                :empty-icon="icons.mdiHeartOutline"
                :full-icon="icons.mdiHeart"
                :half-icon="icons.mdiHeartHalfFull"
                hover
                x-large
                color="red"
                background-color="red"
                @input="updateGifRating"
              />
            </v-item-group>
          </v-row>
        </v-card-actions>
      </v-card>
    </v-row>
  </v-container>
</template>

<script lang="ts">
import { Vue, Component } from 'vue-property-decorator';
import {
  mdiMagnify, mdiTrashCan, mdiDownload, mdiHeartOutline, mdiHeart, mdiHeartHalfFull, mdiEmoticonOutline,
} from '@mdi/js';
import { getTranslateGifFromGiphy, getBlob } from '@/request';
import { saveAs } from 'file-saver';
// @ts-ignore
import { Picker } from 'emoji-mart-vue';

@Component({
  components: {
    Picker,
  },
})
export default class Translate extends Vue {
  icons = {
    mdiMagnify,
    mdiTrashCan,
    mdiDownload,
    mdiHeartOutline,
    mdiHeart,
    mdiHeartHalfFull,
    mdiEmoticonOutline,
  }

  rating: number = 0;

  hasError: boolean = false

  wasSuccessful: boolean = false

  weirdnessSlider: number = 5;

  translatedGif: string = '';

  translatedGifPreview: string = '';

  translatedGifId: string = '';

  isInputValid: boolean = false;

  inputValue: string = '';

  showPicker: boolean = false;

  addEmoji(payload: any) {
    if (this.inputValue.length < 24) { this.inputValue += `${payload.native}`; }
  }

  openEmojiMart() {
    this.showPicker = true;
  }

  clearTextField() {
    this.translatedGif = '';
    this.translatedGifId = '';
    this.hasError = false;
    this.wasSuccessful = false;
    this.rating = 0;
  }

  async getRating() {
    const gifData = await this.$store.dispatch('getGifData', `ggid-${this.translatedGifId}`);
    this.rating = gifData ? this.rating = gifData.rating : 0;
  }

  async getTranslatedGif() {
    try {
      const result = await getTranslateGifFromGiphy(this.inputValue, this.weirdnessSlider);
      this.translatedGifId = result.id;
      this.translatedGifPreview = result.images.fixed_width.webp;
      this.translatedGif = result.images.original.webp || result.images.original.url;
      await this.getRating();
      this.hasError = false;
      this.wasSuccessful = true;
    } catch {
      this.hasError = true;
      this.wasSuccessful = false;
    }
  }

  async removeGif() {
    this.rating = 0;
    await this.$store.dispatch('removeGifData', `ggid-${this.translatedGifId}`);
    this.$store.commit('setGifCount', this.$store.state.gifCount - 1);
  }

  // eslint-disable-next-line class-methods-use-this
  async getBlobForGif(imageData: string) {
    const previewGifBlob = await getBlob(imageData);
    return previewGifBlob;
  }

  async updateGifRating(rating: number) {
    if (this.$store.state.fullImageMode) {
      await this.$store.dispatch('setGifData', {
        rating,
        key: `ggid-${this.translatedGifId}`,
        image: await this.getBlobForGif(this.translatedGif),
        preview: await this.getBlobForGif(this.translatedGifPreview),
      });
    } else {
      await this.$store.dispatch('setGifData', {
        rating,
        key: `ggid-${this.translatedGifId}`,
        image: this.translatedGif,
        preview: this.translatedGifPreview,
      });
    }
    const currentCount = await this.$store.dispatch('getGifCount');
    this.$store.commit('setGifCount', currentCount);
  }

  saveImage(): void {
    saveAs(this.translatedGif, `ggid-${this.translatedGifId}`);
  }
}
</script>

<style scoped>
#trashIcon:hover {
  color: red;
}
#downloadIcon:hover {
  color: green;
}
</style>
