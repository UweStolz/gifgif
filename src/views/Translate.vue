<template>
  <TranslateView />
</template>

<script lang="ts">
import { Component, Vue } from 'vue-property-decorator';
import TranslateView from '../components/Translate/index.vue';

@Component({
  components: {
    TranslateView,
  },
})
export default class Translate extends Vue { }
</script>
