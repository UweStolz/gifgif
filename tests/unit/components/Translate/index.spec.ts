import Translate from '@/components/Translate/index.vue';
import shallow from '../../../helper';

describe('Translate.vue', () => {
  it('renders properly', () => {
    const wrapper = shallow(Translate);
    expect(wrapper).toMatchSnapshot();
  });
});
