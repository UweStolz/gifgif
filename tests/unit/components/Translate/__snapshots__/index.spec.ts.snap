// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Translate.vue renders properly 1`] = `
<v-container-stub tag="div" fluid="true">
  <v-row-stub tag="div" align="start" justify="center">
    <v-card-stub loaderheight="4" tag="div" width="80%" tile="true" flat="true">
      <v-slider-stub errorcount="1" errormessages="" messages="" rules="" successmessages="" value="5" backgroundcolor="" label="Weirdness factor" loaderheight="2" max="10" min="1" step="1" thumbcolor="red" thumblabel="always" thumbsize="32" ticklabels="" ticks="always" ticksize="2" trackcolor="black" trackfillcolor="red darken-3" style="margin-top: 30px;"></v-slider-stub>
      <v-form-stub>
        <v-text-field-stub color="black" errorcount="1" errormessages="" messages="" rules="" successmessages="" value="" backgroundcolor="" prependicon="M9.5,3A6.5,6.5 0 0,1 16,9.5C16,11.11 15.41,12.59 14.44,13.73L14.71,14H15.5L20.5,19L19,20.5L14,15.5V14.71L13.73,14.44C12.59,15.41 11.11,16 9.5,16A6.5,6.5 0 0,1 3,9.5A6.5,6.5 0 0,1 9.5,3M9.5,5C7,5 5,7 5,9.5C5,12 7,14 9.5,14C12,14 14,12 14,9.5C14,7 12,5 9.5,5Z" loaderheight="2" clearable="true" clearicon="$clear" counter="true" placeholder="Enter a search phrase" prependinnericon="M12,17.5C14.33,17.5 16.3,16.04 17.11,14H6.89C7.69,16.04 9.67,17.5 12,17.5M8.5,11A1.5,1.5 0 0,0 10,9.5A1.5,1.5 0 0,0 8.5,8A1.5,1.5 0 0,0 7,9.5A1.5,1.5 0 0,0 8.5,11M15.5,11A1.5,1.5 0 0,0 17,9.5A1.5,1.5 0 0,0 15.5,8A1.5,1.5 0 0,0 14,9.5A1.5,1.5 0 0,0 15.5,11M12,20A8,8 0 0,1 4,12A8,8 0 0,1 12,4A8,8 0 0,1 20,12A8,8 0 0,1 12,20M12,2C6.47,2 2,6.5 2,12A10,10 0 0,0 12,22A10,10 0 0,0 22,12A10,10 0 0,0 12,2Z" type="text" maxlength="25"></v-text-field-stub>
      </v-form-stub>
      <!---->
      <v-row-stub tag="div" align="start" justify="center">
        <!---->
      </v-row-stub>
      <!---->
    </v-card-stub>
  </v-row-stub>
</v-container-stub>
`;
